File: MainRiceCooker.dart
import 'dart:async';
import 'dart:io';

class RiceCooker {
  int? quantityOfRice;
  int? quantityOfWater;
  int temperature = 100;
  int timer = 30;
  String currentStatus = 'Idle';
  int currentTemperature = 0;
  Timer? currentTimer;
  bool isCooking = false;

  void setupCooking(
      [int? riceQuantity, int? waterQuantity, int? temperature, int? timer]) {
    this.quantityOfRice = riceQuantity;
    this.quantityOfWater = waterQuantity;
    this.currentTemperature = temperature ?? this.temperature;
    this.timer = timer ?? this.timer;
    currentStatus = 'Setup completed. Ready to start cooking.';
    print('Setup completed. Ready to start cooking.');
  }

  void startCooking() {
    if (quantityOfWater == null || quantityOfWater == 0) {
      print(
          'Error: Insufficient water. Please add water before starting cooking.');
      return;
    }

    currentTimer?.cancel();
    currentTimer = Timer(
      Duration(minutes: timer),
      () {
        currentStatus = 'Cooking complete. Starting warm mode.';
        isCooking = false;
        warmMode();
      },
    );

    isCooking = true;
    currentStatus = 'Cooking started. Timer set for $timer minutes.';
    print('Cooking started. Timer set for $timer minutes.');
  }

  void warmMode() {
    if (quantityOfWater == null || quantityOfWater == 0) {
      print(
          'Error: Insufficient water. Please add water before starting Warm mode.');
      return;
    }
    currentTemperature = 60;
    currentStatus = 'Warm mode activated. Temperature set to 60°C.';
    print('Warm mode activated. Temperature set to 60°C.');
  }

  void pauseCooking() {
    if (isCooking) {
      currentTimer?.cancel();
      currentTemperature = 0;
      isCooking = false;
      currentStatus = 'Cooking paused.';
      print('Cooking paused.');
    } else {
      print('No active cooking process to pause.');
    }
  }

  void resumeCooking() {
    if (isCooking) {
      print('Already cooking. Cannot resume.');
    } else {
      currentTimer = Timer(
        Duration(minutes: timer),
        () {
          currentStatus = 'Cooking complete. Starting warm mode.';
          isCooking = false;
          warmMode();
        },
      );
      isCooking = true;
      currentTemperature = temperature;
      currentStatus = 'Cooking resumed. Timer set for $timer minutes.';
      print('Cooking resumed. Timer set for $timer minutes.');
    }
  }

  void checkStatus() {
    print('Status: $currentStatus');
    print('Temperature: $currentTemperature °C');
  }
}

String? readlnOrNull() {
  try {
    return stdin.readLineSync();
  } catch (e) {
    return null;
  }
}

void main() {
  var cooker = RiceCooker();

  print('Welcome to the Rice Cooker CLI!');

  while (true) {
    print('1. Setup Cooking');
    print('2. Start Cooking');
    print('3. Warm Mode');
    print('4. Pause Cooking');
    print('5. Resume Cooking');
    print('6. Check Status');
    print('7. Quit');

    stdout.write('Enter the number of the command: ');
    var userInput = readlnOrNull();
    switch (userInput) {
      case '1':
        try {
          stdout.write('Enter quantity of rice (can be null): ');
          var riceQuantity = int.tryParse(readlnOrNull() ?? '') ?? 0;

          stdout.write('Enter quantity of water (can be null): ');
          var waterQuantity = int.tryParse(readlnOrNull() ?? '') ?? 0;

          stdout.write('Enter temperature (optional, default 100°C): ');
          var temperature = int.tryParse(readlnOrNull() ?? '') ?? 100;

          stdout.write('Enter timer (optional, default 30min): ');
          var timer = int.tryParse(readlnOrNull() ?? '') ?? 30;

          cooker.setupCooking(riceQuantity, waterQuantity, temperature, timer);
        } catch (e) {
          print('Error: Invalid input. Please enter a valid number.');
        }
        break;
      case '2':
        cooker.startCooking();
        break;
      case '3':
        cooker.warmMode();
        break;
      case '4':
        cooker.pauseCooking();
        break;
      case '5':
        cooker.resumeCooking();
        break;
      case '6':
        cooker.checkStatus();
        break;
      case '7':
        print('Exiting Rice Cooker CLI. Goodbye!');
        return;
      default:
        print('Invalid command. Please enter a valid number.');
    }
  }
}

File: MainRiceCooker.go
package main

import (
	"fmt"
	"log"
	"os"
	"time"
)

type RiceCooker struct {
	quantityOfRice     int
	quantityOfWater    int
	temperature        int
	timer              int
	currentStatus      string
	currentTemperature int
	currentTimer       *time.Timer
	isCooking          bool
}

func (rc *RiceCooker) setupCooking(riceQuantity, waterQuantity, temperature, timer int) {
	if riceQuantity <= 0 || waterQuantity <= 0 {
		logError("Quantity of rice and water must be greater than 0 during setup.")
		return
	}

	rc.quantityOfRice = riceQuantity
	rc.quantityOfWater = waterQuantity

	if temperature <= 0 {
		logError("Temperature must be greater than 0 during setup.")
		return
	}

	rc.temperature = temperature

	if timer <= 0 {
		logError("Timer must be greater than 0 during setup.")
		return
	}

	rc.timer = timer

	rc.currentStatus = "Setup completed. Ready to start cooking."
	fmt.Println("Setup completed. Ready to start cooking.")
}

func (rc *RiceCooker) startCooking() {
	if rc.quantityOfWater <= 0 {
		logError("Insufficient water. Please add water before starting cooking.")
		return
	}

	if rc.currentTimer != nil {
		rc.currentTimer.Stop()
	}

	rc.currentTimer = time.NewTimer(time.Duration(rc.timer) * time.Minute)

	go func() {
		<-rc.currentTimer.C
		rc.currentStatus = "Cooking complete. Starting warm mode."
		rc.isCooking = false
		rc.warmMode()
	}()

	rc.isCooking = true
	rc.currentStatus = fmt.Sprintf("Cooking started. Timer set for %d minutes.", rc.timer)
	fmt.Printf("Cooking started. Timer set for %d minutes.\n", rc.timer)
}

func (rc *RiceCooker) warmMode() {
	if rc.quantityOfWater <= 0 {
		logError("Insufficient water. Please add water before starting Warm mode.")
		return
	}
	rc.currentTemperature = 60
	rc.currentStatus = "Warm mode activated. Temperature set to 60°C."
	fmt.Println("Warm mode activated. Temperature set to 60°C.")
}

func (rc *RiceCooker) pauseCooking() {
	if rc.isCooking {
		rc.currentTimer.Stop()
		rc.currentTemperature = 0
		rc.isCooking = false
		rc.currentStatus = "Cooking paused."
		fmt.Println("Cooking paused.")
	} else {
		fmt.Println("No active cooking process to pause.")
	}
}

func (rc *RiceCooker) resumeCooking() {
	if rc.isCooking {
		fmt.Println("Already cooking. Cannot resume.")
	} else {
		rc.currentTimer = time.NewTimer(time.Duration(rc.timer) * time.Minute)

		go func() {
			<-rc.currentTimer.C
			rc.currentStatus = "Cooking complete. Starting warm mode."
			rc.isCooking = false
			rc.warmMode()
		}()

		rc.isCooking = true
		rc.currentTemperature = rc.temperature
		rc.currentStatus = fmt.Sprintf("Cooking resumed. Timer set for %d minutes.", rc.timer)
		fmt.Printf("Cooking resumed. Timer set for %d minutes.\n", rc.timer)
	}
}

func (rc *RiceCooker) checkStatus() {
	fmt.Printf("Status: %s\n", rc.currentStatus)
	fmt.Printf("Temperature: %d °C\n", rc.currentTemperature)
}

func logError(message string) {
	logFile, err := os.OpenFile("error_log.txt", os.O_CREATE|os.O_WRONLY|os.O_APPEND, 0666)
	if err != nil {
		fmt.Println("Error opening log file:", err)
		return
	}
	defer logFile.Close()

	log.SetOutput(logFile)
	log.Println("Error:", message)
	fmt.Println("Error:", message)
}

func main() {
	cooker := &RiceCooker{}

	fmt.Println("Welcome to the Rice Cooker CLI!")

	for {
		fmt.Println("1. Setup Cooking")
		fmt.Println("2. Start Cooking")
		fmt.Println("3. Warm Mode")
		fmt.Println("4. Pause Cooking")
		fmt.Println("5. Resume Cooking")
		fmt.Println("6. Check Status")
		fmt.Println("7. Quit")

		fmt.Print("Enter the number of the command: ")
		var userInput string
		fmt.Scanln(&userInput)

		switch userInput {
		case "1":
			var riceQuantity, waterQuantity, temperature, timer int
			fmt.Print("Enter quantity of rice: ")
			_, err := fmt.Scanln(&riceQuantity)
			if err != nil {
				logError("Invalid input. Please enter a valid number.")
				break
			}

			fmt.Print("Enter quantity of water: ")
			_, err = fmt.Scanln(&waterQuantity)
			if err != nil {
				logError("Invalid input. Please enter a valid number.")
				break
			}

			fmt.Print("Enter temperature: ")
			_, err = fmt.Scanln(&temperature)
			if err != nil {
				logError("Invalid input. Please enter a valid number.")
				break
			}

			fmt.Print("Enter timer: ")
			_, err = fmt.Scanln(&timer)
			if err != nil {
				logError("Invalid input. Please enter a valid number.")
				break
			}

			cooker.setupCooking(riceQuantity, waterQuantity, temperature, timer)

		case "2":
			cooker.startCooking()

		case "3":
			cooker.warmMode()

		case "4":
			cooker.pauseCooking()

		case "5":
			cooker.resumeCooking()

		case "6":
			cooker.checkStatus()

		case "7":
			fmt.Println("Exiting Rice Cooker CLI. Goodbye!")
			return

		default:
			fmt.Println("Invalid command. Please enter a valid number.")
		}
	}
}
